W-23176605 oauth metadata policy gr (#1102)
mulesoft/docs-gateway: gateway/1.13/modules/ROOT/pages/policies-included-oauth-protected-resource-metadata.adoc

:page-aliases: policies::policies-included-oauth-protected-resource-metadata.adoc

[width="100%", cols="5,15"]
|===
>s| Policy Name | OAuth 2.0 Protected Resource Metadata
>s|Summary      | Serves OAuth 2.0 Protected Resource Metadata at the well-known endpoint
>s|Category | Security
>s|First Omni Gateway version available | v1.13.0
.2+>.^s| Returned Status Codes
|404 - Request path does not match the well-known endpoint
|405 - HTTP method is not GET
|===

[[summary]]
== Summary

The OAuth Protected Resource Metadata policy serves a JSON document describing the protected resource's OAuth 2.0 metadata at `/.well-known/oauth-protected-resource` as defined by https://www.rfc-editor.org/rfc/rfc9728[RFC 9728^]. The metadata document is defined by the policy. The policy doesn't fetch metadata from an external source.

This policy enables authorization servers and client applications to discover important OAuth 2.0 configuration details about your protected resource, such as:

* Which authorization server tokens are accepted
* What scopes are available for authorization
* How to send bearer tokens (header, body, or query)
* Whether DPoP or mTLS token binding is required

> * URLs for resource documentation, policy, and terms of service 

Reviewer: ```suggestion
* What URLs are available for resource documentation, policy, and terms of service 
```

include::partial$policy-title-headers.adoc[tag=ui]

[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Parameter | Required or Optional | Default Value | Description

| Resource Identifier
| Required
|
| The protected resource identifier (RFC 8707). Must be an `https` URL without a fragment. This value is returned as the `resource` field in the metadata response and is used by clients to validate the response matches the resource they intended to query. +
Example: `"https://api.example.com/v1"`

| Authorization Servers
| Optional
| (omitted from response)
| Array of OAuth 2.0 authorization server issuer identifier URLs (per RFC 8414) whose tokens are accepted by this protected resource. Enter multiple values separated by commas or on separate lines. +
Example: `["https://auth.example.com"]`

| JWK Set URI
| Optional
| (omitted from response)
| URL of the resource's JWK Set document containing its public keys for signing resource responses. Must use `https` scheme. +
Example: `"https://api.example.com/.well-known/jwks.json"`

| Scopes Supported
| Optional (recommended by RFC 9728)
| (omitted from response)

> | Array of default OAuth 2.0 scope values that can be used to request access to this resource. This value is returned when the well-known path has no suffix or when the suffix does not match any entry in Resource Scopes Overrides. Enter multiple values separated by commas or on separate lines. +

Reviewer: ```suggestion
| Array of default OAuth 2.0 scope values that can be used to request access to this resource. This value is returned when the well-known path has no suffix or when the suffix doesn't match any entry in Resource Scopes Overrides. Enter multiple values separated by commas or on separate lines. +
```

* A `GET` request to the default endpoint, `/.well-known/oauth-protected-resource`, returns the top-level scopes:
+
[source,json]
----
{
  "resource": "https://api.example.com",
  "authorization_servers": ["https://auth.example.com"],
  "scopes_supported": ["read", "write"]
}
----
+
* A `GET` request to `/.well-known/oauth-protected-resource/users` matches the `users` override and returns its scopes:
+
[source,json]
----
{
  "resource": "https://api.example.com",
  "authorization_servers": ["https://auth.example.com"],
  "scopes_supported": ["users:read", "users:write", "users:delete"]
}
----
+
* A `GET` request to an unconfigured suffix, such as `/.well-known/oauth-protected-resource/billing`, falls back to the top-level scopes:
+
[source,json]
----
{
  "resource": "https://api.example.com",
  "authorization_servers": ["https://auth.example.com"],
  "scopes_supported": ["read", "write"]
}
----

[[edge-cases]]
== Edge Cases

* *Non-matching paths:* Any request whose path doesn't contain `/.well-known/oauth-protected-resource` receives `404 Not Found`. The policy is intended to be route-scoped to the well-known path.

> * *Trailing slash:* `/.well-known/oauth-protected-resource/` (with trailing slash) is treated as a match with an empty suffix. Since an empty string won't match any `resourceScopesOverrides` entry, the top-level `scopesSupported` is returned.

Reviewer: ```suggestion
* *Trailing slash:* `/.well-known/oauth-protected-resource/` (with trailing slash) is treated as a match with an empty suffix. Since an empty string won't match any `resourceScopesOverrides` entries, the top-level `scopesSupported` is returned.
```
